pkg/validator/shoot_validator.go

```diff
@@ -0,0 +1,159 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validator
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	gcpvalidation "github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp/validation"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"github.com/gardener/gardener/pkg/apis/core"
+)
+
+var (
+	specPath = field.NewPath("spec")
+
+	providerConfigPath = specPath.Child("providerConfig")
+	networkPath        = specPath.Child("networking")
+	providerPath       = specPath.Child("provider")
+
+	infrastructureConfigPath = providerPath.Child("infrastructureConfig")
+	controlPlaneConfigPath   = providerPath.Child("controlPlaneConfig")
+	workersPath              = providerPath.Child("workers")
+)
+
+type validationContext struct {
+	shoot                *core.Shoot
+	infrastructureConfig *gcp.InfrastructureConfig
+	controlPlaneConfig   *gcp.ControlPlaneConfig
+	cloudProfile         *gardencorev1beta1.CloudProfile
+	cloudProfileConfig   *gcp.CloudProfileConfig
+}
+
+// getAllowedRegionZonesFromCloudProfile fetches the set of allowed zones from the Cloud Profile.
+func getAllowedRegionZonesFromCloudProfile(shoot *core.Shoot, cloudProfile *gardencorev1beta1.CloudProfile) sets.String {
+	shootRegion := shoot.Spec.Region
+	for _, region := range cloudProfile.Spec.Regions {
+		if region.Name == shootRegion {
+			gcpZones := sets.NewString()
+			for _, gcpZone := range region.Zones {
+				gcpZones.Insert(gcpZone.Name)
+			}
+			return gcpZones
+		}
+	}
+	return nil
+}
+
+func (v *Shoot) validateShootCreate(ctx context.Context, shoot *core.Shoot) (field.ErrorList, error) {
+	valContext, err := newValidationContext(ctx, v.decoder, v.client, shoot)
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		allErrors    = field.ErrorList{}
+		allowedZones = getAllowedRegionZonesFromCloudProfile(shoot, valContext.cloudProfile)
+	)
+
+	allErrors = append(allErrors, gcpvalidation.ValidateNetworking(shoot.Spec.Networking, networkPath)...)
+	allErrors = append(allErrors, gcpvalidation.ValidateInfrastructureConfig(valContext.infrastructureConfig, shoot.Spec.Networking.Nodes, shoot.Spec.Networking.Pods, shoot.Spec.Networking.Services, infrastructureConfigPath)...)
+	allErrors = append(allErrors, gcpvalidation.ValidateControlPlaneConfig(valContext.controlPlaneConfig, allowedZones, controlPlaneConfigPath)...)
+	allErrors = append(allErrors, gcpvalidation.ValidateWorkers(shoot.Spec.Provider.Workers, workersPath)...)
+
+	return allErrors, allErrors.ToAggregate()
```

Comment:
Can you extract these lines to a separate func `validateShoot` and call it from both `validateShootCreate` and `validateShootUpdate`. This way the `validationcontext` for the current shoot is only created once in case of an update.

Reply:
can do, but then you will have another function?  having a duplicate validate shoot is a bit confusing to me.

Reply:
We can have two functions, one for `create` and one for `update` and then one `validate` function which bundles the validation logic which must be executed for both events. So I don't see any duplication?

Reply:
there is no other code being called in validateShootCreation after the code is moved other than validateShoot.